Прохожу спринт 15

EntityVadim opened the pull request:
Дорогой ревью, прошу тебя внести комментарии по коду, может, где-то, что-то стоит поменять, улучшить и т.д. или ссылки на полезные статьи, как я могу изменить свой код, может какие-то лайфхаки крутые подскажешь.
Буду тебе очень сильно в этом благодарен!

EntityVadim committed:
- Сделал более читаемый код.

EntityVadim committed:
- Сделал более читаемым код.

EntityVadim committed:
- Создал две коллекции: эмоджи и цвет.

EntityVadim committed:
- Настроил правильно расположение коллекции.

EntityVadim committed:
- Обновил цвета.

EntityVadim committed:
- Настроил правильно коллекцию.

EntityVadim committed:
- Сделал более читаемы код.

EntityVadim committed:
- Мелкие доработки по коду.

EntityVadim committed:
- Выполнил задние:
  1. добавлены Entity для хранения Tracker, TrackerCategory, TrackerRecord.

EntityVadim committed:
- Пытаюсь добавить скролвью в экран "Создания привычки";
- Пытаюсь сохранить данные в CoreData.

EntityVadim committed:
- Исправил проблему с сохранением категорий в CoreData.

EntityVadim committed:
- Мелкие исправления по коду.

EntityVadim committed:
- Пытаюсь понять как исправить ошибку Value of type 'TrackerDataManager' has no member 'categories'.

EntityVadim committed:
- Решаю проблему с сохранение CoreDate.

EntityVadim committed:
- Исправил ошибку с CoreData, но теперь появилась ошибка с отображением категорий и трекеров

EntityVadim committed:
- Уменьшил код в ширину для лучший читаемости

EntityVadim committed:
- При создании привычки трекер отображается.
(Возникли ошибки: 1. Дублируется трекер; 2. Не отмечается завершенным.)

EntityVadim committed:
- Устранил ошибку по дублирования трекера при его создании.

EntityVadim committed:
- Решил проблему отображения трекера по расписанию.

EntityVadim committed:
- Исправил баг с отображением трекеров.

EntityVadim committed:
- Исправил баг с завершением трекера.

EntityVadim committed:
- Оптимизировал код.

EntityVadim committed:
- Настроил правильно скролвью.

LizaIlinykh commented on Tracker/Trackers/TrackerCategory/ExtensionTrackerCategory.swift:
```diff
@@ -0,0 +1,77 @@
+        _ tableView: UITableView,
+        cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+            guard let cell = tableView.dequeueReusableCell(
+                withIdentifier: "CustomCategoryCell",
```
Идентификаторы лучше хранить в константах , например в самой ячейке , со строками легче допустить ошибку

LizaIlinykh commented on Tracker/Trackers/TrackerCategory/TrackerCategoryViewController.swift:
```diff
@@ -13,10 +13,27 @@ final class TrackerCategoryViewController: UIViewController {
     
     // MARK: - Private Properties
     
+    private let trackerCategoryStore = TrackerCategoryStore()
```
лучше внедрять зависимости извне
вот статья про внедрение зависимостей: https://swiftblog.org/vnedrenie-zavisimostej/
также стоит закрыть сторадж протоколом

LizaIlinykh commented on Tracker/Trackers/TrackerCategory/TrackerCategoryViewController.swift:
```diff
@@ -101,21 +107,27 @@ final class TrackerCategoryViewController: UIViewController {
     private func loadSelectedCategory() {
-        selectedCategory = UserDefaults.standard.string(forKey: "selectedCategory")
-        tableView.reloadData()
+        if let title = UserDefaults.standard.string(forKey: "selectedCategory") {
```
рекомендую ключ вынести в константу

LizaIlinykh replied:
здесь лучше использовать конструкцию guard

LizaIlinykh commented on Tracker/Trackers/TrackerCreation/ExtensionTrackerCreation.swift:
```diff
@@ -0,0 +1,75 @@
+            if collectionView == emojiCollectionView {
+                let cell = collectionView.dequeueReusableCell(
+                    withReuseIdentifier: EmojiCell.reuseIdentifier,
+                    for: indexPath) as! EmojiCell
```
рекомендую избегать принудительных операций

LizaIlinykh commented on Tracker/Trackers/TrackerCreation/ExtensionTrackerCreation.swift:
```diff
@@ -0,0 +1,75 @@
+            } else {
+                let cell = collectionView.dequeueReusableCell(
+                    withReuseIdentifier: ColorCell.reuseIdentifier,
+                    for: indexPath) as! ColorCell
```
рекомендую избегать принудительных операций

LizaIlinykh commented on Tracker/Trackers/TrackerSchedule/TrackerScheduleViewController.swift:
```diff
@@ -23,12 +23,21 @@ enum WeekDay: String, CaseIterable {
     var daySelectionHandler: (([WeekDay]) -> Void)?
     
-    // MARK: - UI Elements
+    lazy var tableView: UITableView = {
```
давай сделаем private(set), чтобы обезопасить вьюху от изменений извне

LizaIlinykh commented on Tracker/Trackers/TrackerViewController/ExtensionTracker.swift:
```diff
@@ -0,0 +1,121 @@
+extension TrackerViewController:
+    UICollectionViewDataSource,
+    UICollectionViewDelegate,
+    UICollectionViewDelegateFlowLayout {
```
лучше не подписываться в одном расширении сразу не несколько протоколов, лучше разделить на несколько и реализовывать методы каждого протокола в отдельном расширении. Это сделает код более читабельным

LizaIlinykh commented on Tracker/Trackers/TrackerViewController/ExtensionTracker.swift:
```diff
@@ -0,0 +1,121 @@
+    
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
+        let sectionCount = dataManager.categories.count
+        return sectionCount
```
рекомендую сокращать код, там где это возможно

например здесь:

```
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        dataManager.categories.count
    }
```

LizaIlinykh commented on Tracker/Trackers/TrackerViewController/ExtensionTracker.swift:
```diff
@@ -0,0 +1,121 @@
+            cell.configure(
+                with: tracker,
+                completedTrackers: completedTrackers,
+                dataManager: dataManager,
```
внедрять dataManager в ячейку не стоит, ячейка должна содержать как можно меньше бизнес-логики. 
если она хочет сообщить о том, что произошло пользовательское действие (например, нажатие кнопки), она должна просто сообщить об этом делегату, а делегат сам должен решить как обработать это событие.
уменьшние количества бизнес-логики в вьюхах делает их более переиспользуемыми

LizaIlinykh commented on Tracker/Trackers/TrackerViewController/ExtensionTracker.swift:
```diff
@@ -0,0 +1,121 @@
+            let completedTrackers = dataManager.completedTrackers.filter { $0.trackerId == tracker.id }
+            cell.configure(
+                with: tracker,
+                completedTrackers: completedTrackers,
```
completedTrackers - это тоже бизнес логика, не рекомендую ее передавать ячейке

LizaIlinykh commented on Tracker/Trackers/TrackerViewController/TrackerViewController.swift:
```diff
@@ -11,17 +11,19 @@ import UIKit
     
-    private var dataManager = TrackerDataManager()
-    private var selectedDate: Date = Date()
+    let dataManager = TrackerDataManager.shared
```
давай поставим private(set) модификатор и внедрим эту зависимость через инициализатор

LizaIlinykh commented on Tracker/Trackers/TrackerViewController/TrackerViewController.swift:
```diff
@@ -11,17 +11,19 @@ import UIKit
+    var selectedDate: Date = Date()
     
-    private let dateFormatter: DateFormatter = {
+    let dateFormatter: DateFormatter = {
```
dateFormatter нужен здесь только TrackerDataManager , давай там его и инициализируем и сделаем приватным свойством

LizaIlinykh commented on Tracker/Trackers/TrackerViewController/ExtensionTracker.swift:
```diff
@@ -0,0 +1,121 @@
+            }
+            let tracker = trackers[indexPath.item]
+            let completedTrackers = dataManager.completedTrackers.filter { $0.trackerId == tracker.id }
+            cell.configure(
```
давай в cell будем передавать только модель, которая будет содержать готовые данные для отображения
она будет выглядеть так
```
struct TrackerCellViewModel {
let title: String // для названия трекера
let subtitle: String // количество дней
let emoji: String // эмоджи
let isCompleted: Bool // как нужно отобразить кнопку
let color: String // цвет трекера
}
```

это то что должна получить ячейка для отрисовки

LizaIlinykh commented on Tracker/Trackers/TrackersDataStorage/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,53 @@
+        let request = NSFetchRequest<TrackerCategoryCoreData>(entityName: "TrackerCategoryCoreData")
+        do {
+            let authors = try context.fetch(request)
+            authors.forEach { categories.append(TrackerCategory(title: $0.title!, trackers: [])) }
```
рекомендую избегать принудительных операций

EntityVadim committed:
- Избавился от принудительных операций.

EntityVadim committed:
- Использовал конструкцию guard в private func loadSelectedCategory();
- Ключ forKey: "selectedCategory" вынес в константу;
- Идентификатор withIdentifier: "CustomCategoryCell" вынес в константу.

EntityVadim committed:
- Сократил код в func numberOfSections;
- Ключ "isOnboardingCompleted" вынес в константу;
- Идентификатор "TrackerCell", "TrackerSectionHeader" и "cell"  вынес в константы.
- Разделил расширения.

EntityVadim committed:
- Исправил критическое замечание:
"При перезапуске приложения не сохраняется информация о выполненности трекера"

EntityVadim committed:
- Правильно настроил "Нерегулярные события".
